### PR TITLE
make "open terminal here" context menu work for directory background

### LINKED ISSUE #6414: Re-enable the "open terminal here" in the background of a directory
This is going to regress when I submit the "fix" for #6245.

You currently can't use the "Open Windows Terminal Here" shortcut when clicking on the background of a directory, due to an internal bug, MSFT:24623699. This is fixed in the latest OS builds, but hasn't been brought downlevel quite yet.

This issue is tracking the work to re-enable that functionality, once the internal bug is fixed.

### PR DESCRIPTION
**Pull request by hereafter**
This commit makes "Open in Windows Terminal" Context menu work again for
directory background even on system that OS fix is not applied.

This is a fallback solution to OS fixes mentioned in #6414.
While OS fix is on its way, we need a fallback that works on existing OS
versions.

The approach to this is: when no item is selected (nullptr for
IShellItemArray*), we use shell api to query the path of current active
Explorer window. A special case is handled for Windows Desktop. Once
we are able to obtain the path, we launch Windows Terminal with it.

## Validation Steps Performed
1. Right click on desktop to bring up the Context menu, pick "Open in
   Windows Terminal", verify that a terminal is opened with correct
   initial path.

2. Open a few File Explorer windows, pick any window, navigate to a
   folder, click on "Background" to bring up the context menu, click
   "Open in Windows Terminal" verify that a terminal is opened with
   correct initial path.

Closes #6414

### CODE REVIEW

**Comment by DHowett** [comment]
WOW! This is clever!

File: .github/actions/spell-check/expect/a129ff14ec985d6b7bf09e296a831c955d41040b.txt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+checkboxes
```
Comment (carlos-zamora) [inline .github/actions/spell-check/expect/a129ff14ec985d6b7bf09e296a831c955d41040b.txt:1]: was this a file that git added accidentally?

File: src/cascadia/ShellExtension/OpenTerminalHere.cpp [hunk 2]
```diff
@@ -117,14 +118,29 @@ static std::wstring _getExePath()
...
-    DWORD count;
-    psiItemArray->GetCount(&count);
+    wil::unique_cotaskmem_string pszName;
 
-    winrt::com_ptr<IShellItem> psi;
-    RETURN_IF_FAILED(psiItemArray->GetItemAt(0, psi.put()));
+    if (psiItemArray == nullptr)
+    {
+        // get the current path from explorer.exe
+        auto path = this->GetPathFromExplorer();
```
Comment (carlos-zamora) [inline src/cascadia/ShellExtension/OpenTerminalHere.cpp:126]:
```suggestion
        const auto path = this->GetPathFromExplorer();
```

File: src/cascadia/ShellExtension/OpenTerminalHere.h [hunk 3]
```diff
@@ -46,6 +46,9 @@ struct __declspec(uuid("9f156763-7844-4dc4-b2b1-901f640f5155"))
     STDMETHODIMP GetCanonicalName(GUID* pguidCommandName);
     STDMETHODIMP EnumSubCommands(IEnumExplorerCommand** ppEnum);
 #pragma endregion
+
+private:
+    std::wstring GetPathFromExplorer() const;
```
Comment (carlos-zamora) [inline src/cascadia/ShellExtension/OpenTerminalHere.h:51]:
```suggestion
    std::wstring _GetPathFromExplorer() const;
```
Generally we denote private functions with an `_`

**Review by carlos-zamora — CHANGES_REQUESTED** [review]
Requesting some minor changes. This is awesome though, thanks!

File: .github/actions/spell-check/expect/a129ff14ec985d6b7bf09e296a831c955d41040b.txt (hunk 1 above)
Comment (hereafter) [reply .github/actions/spell-check/expect/a129ff14ec985d6b7bf09e296a831c955d41040b.txt:1]: This text file is added with the scripted provided by the bot reply above. 
it helps the github recognize new names to reduce the spell checking warnings.

File: src/cascadia/TerminalConnection/TerminalConnection.vcxproj [hunk 4]
```diff
@@ -8,6 +8,9 @@
     <SubSystem>Console</SubSystem>
     <OpenConsoleUniversalApp>true</OpenConsoleUniversalApp>
   </PropertyGroup>
+  <PropertyGroup Label="Vcpkg">
+    <VcpkgEnabled>false</VcpkgEnabled>
+  </PropertyGroup>
```
Comment (zadjii-msft) [inline src/cascadia/TerminalConnection/TerminalConnection.vcxproj:13]: This change seems unrelated? Can we revert it?

File: src/cascadia/ShellExtension/pch.cpp [hunk 5]
```diff
@@ -2,3 +2,5 @@
 // Licensed under the MIT license.
 
 #include "pch.h"
+
+#pragma comment(lib, "User32.lib")
```
Comment (zadjii-msft) [inline src/cascadia/ShellExtension/pch.cpp:6]: Can we add this in the `.vcxproj` file instead? None of our other projects use the pragma style of including libs, and I wouldn't want someone to get confused in the future. You should be able to add it with 

```xml

  <ItemDefinitionGroup>
    <Link>
      <AdditionalDependencies>User32.lib;%(AdditionalDependencies)</AdditionalDependencies>
    </Link>
  </ItemDefinitionGroup>
```

File: src/cascadia/ShellExtension/OpenTerminalHere.cpp [hunk 6]
```diff
@@ -214,3 +230,93 @@ HRESULT OpenTerminalHere::EnumSubCommands(IEnumExplorerCommand** ppEnum)
...
+    variant.vt = VT_I4;
+
+    com_ptr<IWebBrowserApp> browser;
+    // look for correct explorer window
+    for (variant.intVal = 0;
+         shell->Item(variant, disp.put()) == S_OK;
+         variant.intVal++)
+    {
+        com_ptr<IWebBrowserApp> tmp;
+        if (FAILED(disp->QueryInterface(browser.put())))
```
Comment (zadjii-msft) [inline src/cascadia/ShellExtension/OpenTerminalHere.cpp:286]: Is this supposed to be 
```suggestion
        if (FAILED(disp->QueryInterface(tmp.put())))
```
?

Comment (zadjii-msft) [reply src/cascadia/ShellExtension/OpenTerminalHere.cpp:286]: (Otherwise, it looks like `tmp` is null when it's used below)

File: src/cascadia/ShellExtension/OpenTerminalHere.cpp [hunk 7]
```diff
@@ -214,3 +230,93 @@ HRESULT OpenTerminalHere::EnumSubCommands(IEnumExplorerCommand** ppEnum)
...
+        {
+            disp = nullptr;
```
Comment (zadjii-msft) [inline src/cascadia/ShellExtension/OpenTerminalHere.cpp:288]: Do we really need to be manually resetting `disp` to `nullptr` each time? Won't it get cleaned up correctly just by virtue of being a `com_ptr`?

File: src/cascadia/ShellExtension/OpenTerminalHere.cpp [hunk 8]
```diff
@@ -214,3 +230,93 @@ HRESULT OpenTerminalHere::EnumSubCommands(IEnumExplorerCommand** ppEnum)
...
+            continue;
+        }
+
+        HWND tmpHWND = NULL;
+        hr = tmp->get_HWND(reinterpret_cast<SHANDLE_PTR*>(&tmpHWND));
```
Comment (zadjii-msft) [inline src/cascadia/ShellExtension/OpenTerminalHere.cpp:293]: Can `get_HWND` fail? It doesn't look like we're checking the `hr` here.

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
Okay, I love how clever this is. I've got a bunch of basically nits, but largely this seems good to me. Thanks!

File: src/cascadia/TerminalConnection/TerminalConnection.vcxproj (hunk 4 above)
Comment (hereafter) [reply src/cascadia/TerminalConnection/TerminalConnection.vcxproj:13]: ok

File: src/cascadia/ShellExtension/pch.cpp (hunk 5 above)
Comment (hereafter) [reply src/cascadia/ShellExtension/pch.cpp:6]: alright

File: src/cascadia/ShellExtension/OpenTerminalHere.cpp (hunk 7 above)
Comment (hereafter) [reply src/cascadia/ShellExtension/OpenTerminalHere.cpp:288]: when get_HWND is failed hwnd is nullptr, it will fail the next == comparison,  it will not affect the algorithm

File: src/cascadia/ShellExtension/OpenTerminalHere.cpp (hunk 8 above)
Comment (hereafter) [reply src/cascadia/ShellExtension/OpenTerminalHere.cpp:293]: when get_HWND is failed hwnd is nullptr, it will fail the next == comparison, it will not affect the algorithm

File: src/cascadia/ShellExtension/OpenTerminalHere.cpp (hunk 7 above)
Comment (hereafter) [reply src/cascadia/ShellExtension/OpenTerminalHere.cpp:288]: you are right this is bug. 
disp no need to set to nullptr, i traced the source code, the existing pointer got released in operator =

File: src/cascadia/ShellExtension/OpenTerminalHere.cpp (hunk 6 above)
Comment (hereafter) [reply src/cascadia/ShellExtension/OpenTerminalHere.cpp]: typo here is a bug

**Review by zadjii-msft — APPROVED** [review]

**Review by DHowett — CHANGES_REQUESTED** [review]
You will need to update the other two `appxmanifest` files in the `src/Cascadia/CascadiaPackage` directory. They are used for the Release and Preview versions!

**Comment by fxzxmic** [comment]
Awaiting!!

**Review by carlos-zamora — APPROVED** [review]

**Review by DHowett — APPROVED** [review]
I've chatted with the team who owns explorer shell extensions, and they have conceded that they will accept this workaround. Thanks so much, this is a very clever solution. I love it.

**Comment by hereafter** [comment]
Awesome :) glad it helps.
